js/src/scrollspy.js

```diff
@@ -8,7 +8,7 @@
 import {
   defineJQueryPlugin,
   getElement,
-  getSelectorFromElement,
+  getTargetIdFromElement,
```

Comment:
Thank you @chrisyeh96 for you contribution :smiley:  

Instead of introducing another function to just make the job, you could probably make a minor refactor to

`getElementFromSelector` to check if string starts with `#` and use the `getElementById`. 

After this you can use it inside `getSelectorFromElement` as they use the exactly same code :wink:

Reply:
Unfortunately, I don't think that your suggestion will work with edge cases. For example `#someID > div` is a valid CSS Selector that starts with a `#` but cannot be passed directly to `document.getElementById()`.

I wrote the new function `getTargetIdFromElement()` specifically to be used in cases such as Scrollspy where the Bootstrap documentation explicitly says that the `href` or `data-bs-target` attributes of the element have the format `#someValidHTML5Id`.

The more general `getElementFromSelector()` should still be allowed for generic CSS selectors such as the example above (`#someID > div`).

Reply:
> Unfortunately, I don't think that your suggestion will work with edge cases. For example #someID > div is a valid CSS Selector that starts with a # but cannot be passed directly to document.getElementById().

Maybe this could do the job. Right?

```js

reg = new RegExp(/^#[\w\-]+$/gm)
reg.test(givenId)
```

Reply:
That doesn't work either. As I mentioned in my original issue #34381, any Unicode character that isn't a whitespace is considered a valid HTML5 ID. In particular, numbers, emoji, and non-Latin characters are all valid. These are not matched by your suggested regex.

**My pitch for a separate `getTargetIdFromElement()` method**

In my mind, there are 2 distinct cases that Bootstrap should support:

*Case 1*: Arbitrary CSS Selectors

The Bootstrap docs for components such as [Collapse](https://getbootstrap.com/docs/5.0/components/collapse/#multiple-targets) and [Offcanvas](https://getbootstrap.com/docs/5.0/components/offcanvas/#via-data-attributes) specifically mention supporting arbitrary CSS selectors. These should use the existing `getSelectorFromElement()` / `getElementFromSelector()` methods, which call `document.querySelector()` behind the scenes. By definition of CSS selectors (see official spec below), these components would NOT support IDs that start with digits.

*Case 2*: ID-specific Selectors

The Bootstrap docs for components such as [Scrollspy](https://getbootstrap.com/docs/5.0/components/scrollspy/) and [Carousel](https://getbootstrap.com/docs/5.0/components/carousel/#example) explicitly mention requiring matching an HTML5 ID (see official spec below). These should use the new `getTargetIdFromElement()` method which only matches HTML5 IDs instead of arbitrary CSS selectors. Thus, these components would be able to support IDs that start with digits.

<details><summary>Official Specs for HTML5 IDs and CSS Selectors</summary>

**HTML5 ID**: The [official spec](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:the-id-attribute-2) states
>When specified on HTML elements, the id attribute value must be unique amongst all the IDs in the element's tree and must contain at least one character. The value must not contain any ASCII whitespace. [...] There are no other restrictions on what form an ID can take; in particular, IDs can consist of just digits, start with a digit, start with an underscore, consist of just punctuation, etc.

The method `document.getElementById()` supports all valid HTML5 IDs.

**CSS Selectors**: The [official spec](https://www.w3.org/TR/selectors-4/#id-selector) states
> An ID selector consists of a “number sign” (U+0023, #) immediately followed by the ID value, which must be a CSS identifier

where a CSS identifier is defined ([official spec](https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier)) as follows:
> In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit.

The method `document.querySelector()` only supports valid CSS selectors. For example, `document.querySelector("#3rd")` will raise an exception.

</details>